### PR TITLE
Make nanobind adapter backward compatible

### PR DESCRIPTION
**Pull request by havogt**

### CODE REVIEW

File: include/gridtools/storage/adapter/nanobind_adapter.hpp [hunk 1]
```diff
@@ -24,76 +24,86 @@
 
 namespace gridtools {
     namespace nanobind_sid_adapter_impl_ {
+#if NB_VERSION_MAJOR >= 2
+    using _ssize_t = nanobind::ssize_t;
+#else
+    using _ssize_t = std::size_t;
```
Comment (fthaler) [inline include/gridtools/storage/adapter/nanobind_adapter.hpp:30]: I’d prefer another name here, e.g., `stride_t` or so. Calling an unsigned type similar to the signed `ssize_t` looks a bit misleading to me. Further, apart from the `_impl` namespace and some old `gcl` code, we don’t seem to use names starting with underscore.

File: include/gridtools/storage/adapter/nanobind_adapter.hpp [hunk 2]
```diff
@@ -24,76 +24,86 @@
...
-        auto select_static_stride_value(std::size_t dyn_value) {
-            if constexpr (SpecValue == -1) {
-                return dyn_value;
-            } else {
-                if (SpecValue != dyn_value) {
-                    throw std::invalid_argument("static stride in stride specification doesn't match dynamic stride");
-                }
-                return gridtools::integral_constant<std::size_t, SpecValue>{};
+    inline constexpr nanobind_sid_adapter_impl_::_ssize_t dynamic_size =
+        static_cast<nanobind_sid_adapter_impl_::_ssize_t>(-1);
```
Comment (fthaler) [inline include/gridtools/storage/adapter/nanobind_adapter.hpp:34]: Isn’t the following good enough (and much more readable)?
```suggestion
    inline constexpr _ssize_t dynamic_size = -1;
```

File: include/gridtools/storage/adapter/nanobind_adapter.hpp [hunk 3]
```diff
@@ -24,76 +24,86 @@
...
+    template <std::size_t... Indices>
+    struct dynamic_strides_helper<std::index_sequence<Indices...>> {
+        using type = stride_spec<(void(Indices), dynamic_size)...>;
+    };
+
+    template <std::size_t N>
+    using fully_dynamic_strides = typename dynamic_strides_helper<std::make_index_sequence<N>>::type;
+
+    template <_ssize_t SpecValue>
+    auto select_static_stride_value(std::size_t dyn_value) {
```
Comment (fthaler) [inline include/gridtools/storage/adapter/nanobind_adapter.hpp:53]:
```suggestion
    constexpr auto select_static_stride_value(std::size_t dyn_value) {
```
Probably irrelevant, but just realized that most functions here could probably be `constexpr`.

File: include/gridtools/storage/adapter/nanobind_adapter.hpp [hunk 4]
```diff
@@ -24,76 +24,86 @@
...
             }
+            return gridtools::integral_constant<std::size_t, SpecValue>{};
         }
+    }
 
-        template <nanobind::ssize_t... SpecValues, std::size_t... IndexValues>
-        auto select_static_strides_helper(
-            stride_spec<SpecValues...>, const std::size_t *dyn_values, std::index_sequence<IndexValues...>) {
+    template <_ssize_t... SpecValues, std::size_t... IndexValues>
+    auto select_static_strides_helper(
```
Comment (fthaler) [inline include/gridtools/storage/adapter/nanobind_adapter.hpp:65]:
```suggestion
    constexpr auto select_static_strides_helper(
```

**Review by fthaler — CHANGES_REQUESTED** [review]
Just to note: I haven’t tested with GT4Py, assuming you did.

File: include/gridtools/storage/adapter/nanobind_adapter.hpp (hunk 2 above)
Comment (havogt) [reply include/gridtools/storage/adapter/nanobind_adapter.hpp:34]: that was a relict of moving things around

**Comment by havogt** [comment]
launch jenkins

**Review by fthaler — APPROVED** [review]

**Comment by havogt** [comment]
launch jenkins

**Comment by havogt** [comment]
tested locally, would be good to test it in the gt4py cscs-ci container on daint before merge (but daint is down now)

**Comment by havogt** [comment]
launch jenkins

**Comment by havogt** [comment]
CI passed except ault and dom which are offline

**Comment by havogt** [comment]
test in gt4py ci image seems to work

**Pull request merged**